### ISSUE TITLE
Query with no where clause stalling on foreign table with ~500k rows.

### ISSUE THREAD

**Issue by donnorman** [GeoffMontee/tds_fdw#30]
When performing a query, with or without a where clause, against a foreign table with a clustered index, it's taking so long that it's timing out.  The table in question has only ~560,000 rows of data.
Even if I use a `limit 1` it still takes a long time to return one row of data.
Other similar queries against tables, also with a clustered index, with only ~90,000 rows come back instantly.

Here's an explain verbose/analyze of the query in question.

```
# explain (verbose, analyze) select * from db.table;
                                                                                     QUERY PLAN
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Foreign Scan on db.table  (cost=25.00..564456.00 rows=564431 width=132) (actual time=0.935..148363.049 rows=564431 loops=1)
   Output: id, a, b, c, d, e, f, g, h, i, j
 Planning time: 1798.964 ms
 Execution time: 153531.414 ms
(4 rows)
```

Here's the same explain but using a where clause on the field that the clustered index is based on.

```
# explain (verbose, analyze) select * from db.table where id = 135;
                                                                                     QUERY PLAN
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Foreign Scan on db.table  (cost=25.00..564456.00 rows=564431 width=132) (actual time=1.122..150592.733 rows=1 loops=1)
   Output: id, a, b, c, d, e, f, g, h, i, j
   Filter: (promo_object.id = 135)
   Rows Removed by Filter: 564430
 Planning time: 1262.160 ms
 Execution time: 159899.840 ms
(6 rows)
```

The foreign table was created using the query option instead of table option, if that matters at all how the query against the foreign table performs.

Could this be caused by poorly created indices on the remote table?

I've rebuilt the indices in question and the queries run fine against the sql server.  It's only slow through the fdw.

Is there an option or some other way to force the usage of the index, or is it already using the index?    I can't really tell from the explain.

I had read somewhere to try increasing the `initial block size` setting in the freetds.conf file, but that didn't seem to have any effect on this query.

Any suggestions are appreciated.

Thanks,

-Don

**Comment by GeoffMontee** [comment]
Hi @donnorman 

The problem is that `tds_fdw` doesn't support where clause push-downs yet. See issue #20.

At the moment, if you perform a query like `select * from db.table where id = 135;`, then `tds_fdw` fetches the entire table from the remote server, and then PostgreSQL locally filters the results to find the rows `where id = 135`.

I do want to add where clause push-down some day. Sorry that it isn't available yet.

I also noticed that `Planning time: 1262.160 ms` from your explain output is very long. Is your foreign server running MS SQL Server or Sybase? If it's MS SQL Server, try setting `row_estimate_method` to `showplan_all` for your foreign table. That should speed up planning time.

**Comment by donnorman** [comment]
Thanks @GeoffMontee for the quick response.

I included the `row_estimate_method` in the foreign table definition and it reduced the planning time to `16.616 ms`.

However, even without using a where clause, the query still takes so long that it times out the connection to the sql server.

Here's the explain output for the new table.

```
# explain (verbose, analyze) select * from db.table2;
                                                                                     QUERY PLAN
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Foreign Scan on db.table2  (cost=25.00..564456.00 rows=564431 width=132) (actual time=0.902..149734.124 rows=564431 loops=1)
   Output: id, object_type_id, country_id, literature_code, key_words, created_month_id, created_date, image_count, image_size, publication_date_type_id, publication_date_month_id
 Planning time: 16.616 ms
 Execution time: 157444.218 ms
(4 rows)
```

Here's the results of when I attempt to run the query.

```
server closed the connection unexpectedly
        This probably means the server terminated abnormally
        before or while processing the request.
The connection to the server was lost. Attempting reset: Succeeded.
```

Even though I'm not using a where clause, is this still part of issue #20?

I'll look into timeout settings for freetds.  I don't really need the query to be fast, but I do need it to eventually come back with results.

Great job with the foreign data wrapper, this is helping migrate data out of SQL Server much faster.

**Comment by GeoffMontee** [comment]
Hmm, I'm not sure that your error message actually refers to a FreeTDS timeout. I believe that message usually means that the PostgreSQL server killed the connection to your client.

Does your PostgreSQL error log contain anything that might point to some kind of `tds_fdw` crash or some condition that would cause PostgreSQL to kill your client?

`tds_fdw`'s memory usage isn't great at the moment. There actually seems to be a memory leak in version 1.0.3 (issue #29). You might be hitting some memory limits due to that.

You might also be hitting PostgreSQL's timeout. You might want to try setting [`statement_timeout`](http://www.postgresql.org/docs/9.4/static/runtime-config-client.html) to prevent that.

As you mentioned, setting `timeout` in `freetds.conf` might be useful here too, if the PostgreSQL => SQL Server connection is in fact timing out.

I'm glad you are finding `tds_fdw` useful!

**Comment by donnorman** [comment]
I checked the PostgreSQL log file and this is what I see during the execution of the query.

```
2015-07-08 22:57:41 EDT LOG:  server process (PID 1868) was terminated by signal 9: Killed
2015-07-08 22:57:42 EDT DETAIL:  Failed process was running: select * from db.table ;
2015-07-08 22:57:42 EDT LOG:  terminating any other active server processes
2015-07-08 22:57:42 EDT WARNING:  terminating connection because of crash of another server process
2015-07-08 22:57:42 EDT DETAIL:  The postmaster has commanded this server process to roll back the current transaction and exit, because another server process exited abnormally and possibly corrupted shared memory.
2015-07-08 22:57:42 EDT HINT:  In a moment you should be able to reconnect to the database and repeat your command.
2015-07-08 22:57:42 EDT WARNING:  terminating connection because of crash of another server process
2015-07-08 22:57:42 EDT DETAIL:  The postmaster has commanded this server process to roll back the current transaction and exit, because another server process exited abnormally and possibly corrupted shared memory.
2015-07-08 22:57:42 EDT HINT:  In a moment you should be able to reconnect to the database and repeat your command.
2015-07-08 22:57:43 EDT LOG:  all server processes terminated; reinitializing
2015-07-08 22:57:43 EDT LOG:  database system was interrupted; last known up at 2015-07-08 22:52:02 EDT
2015-07-08 22:57:43 EDT LOG:  database system was not properly shut down; automatic recovery in progress
2015-07-08 22:57:43 EDT LOG:  record with zero length at 0/23AEBB0
2015-07-08 22:57:43 EDT LOG:  redo is not required
2015-07-08 22:57:44 EDT LOG:  database system is ready to accept connections
2015-07-08 22:57:44 EDT LOG:  autovacuum launcher started
```

Just for the heck of it I included the `set memory stats` commands from issue #29 but did not see anything different in the log file versus what is above.

Googling that error seems to point to more a timeout or memory issue with our PostgreSQL installation more than tds_fdw.  I'll try tweaking some settings in the conf and see if the error changes at all.

I am using tds_fdw v1.0.3 by the way, and EnterpriseDB v9.4.

**Comment by GeoffMontee** [comment]
If you are on Linux, PostgreSQL may have been killed by the OOM killer. Does your syslog show anything like that?

**Comment by donnorman** [comment]
Yeah, that is exactly what is going on, thanks.

```
kernel: Out of memory: Kill process 2002 (edb-postgres) score 915 or sacrifice child
kernel: Killed process 2002, UID 500, (edb-postgres) total-vm:6298396kB, anon-rss:3510012kB, file-rss:1884kB
```

It's running on VMware, I'll try feeding it more RAM.

**Comment by donnorman** [comment]
I doubled the RAM and it runs the query just fine, the query even starts returning results in a fraction of the time it used to take for when it returned nothing but the error.

I'll update the title of this issue since it really never had anything to do with indices after all, haha.

**Comment by GeoffMontee** [comment]
Awesome! It's good to hear the query is working now.

Fixing the memory leak is probably my top priority for `tds_fdw` at the moment, since 1.0.3's memory usage seems much worse than 1.0.2. Hopefully I can find time to fix that soon, and then queries shouldn't need as much memory.

**Comment by GeoffMontee** [comment]
Fixed in v1.0.4.
